webserver/src/main.rs

```diff
@@ -0,0 +1,417 @@
+}
+
+#[derive(Debug, Clone, Copy, Default)]
+struct RawStats {
+    draws: usize,
+    wins: usize,
+    losses: usize,
+}
+
+#[derive(Debug, Clone, Default)]
+struct FinalStats {
+    name: String,
+    games: usize,
+    draws: usize,
+    wins: usize,
+    losses: usize,
+    points: f64,
+    pointrate: f64,
+}
+
+async fn stats_with_duration<D: DatabaseReader>(req: Request<D>) -> tide::Result {
+    let duration_str = req.param("duration")?;
+    let duration = humantime::parse_duration(duration_str)?;
+    stats(req, Some(duration.try_into()?)).await
+}
+
+async fn stats<D: DatabaseReader>(
+    req: Request<D>,
+    lookback: Option<time::Duration>,
+) -> tide::Result {
+    let now = OffsetDateTime::now_utc();
+    let range_start = match lookback {
+        None => OffsetDateTime::UNIX_EPOCH,
+        Some(d) => now - d,
+    };
+    let games = req
+        .state()
+        .finished_games(range_start..now)
+        .map_err(anyhow::Error::from)?;
+    let mut player_stats = HashMap::<String, RawStats>::new();
+    let mut team_stats = HashMap::<[String; 2], RawStats>::new();
+
+    for (_, game) in games.into_iter() {
+        let red_team = sort([game.player_red_a, game.player_red_b]);
+        let blue_team = sort([game.player_blue_a, game.player_blue_b]);
+        match game.result.as_str() {
+            "DRAW" => {
+                for p in red_team.iter().chain(blue_team.iter()).cloned() {
+                    player_stats.entry(p).or_default().draws += 1;
+                }
+                for team in [red_team, blue_team].iter() {
+                    team_stats.entry(team.clone()).or_default().draws += 1;
+                }
+            }
+            "VICTORY_RED" => {
+                for p in red_team.iter().cloned() {
+                    player_stats.entry(p).or_default().wins += 1;
+                }
+                for p in blue_team.iter().cloned() {
+                    player_stats.entry(p).or_default().losses += 1;
+                }
+                team_stats.entry(red_team).or_default().wins += 1;
+                team_stats.entry(blue_team).or_default().losses += 1;
+            }
+            "VICTORY_BLUE" => {
+                for p in red_team.iter().cloned() {
+                    player_stats.entry(p).or_default().losses += 1;
+                }
+                for p in blue_team.iter().cloned() {
+                    player_stats.entry(p).or_default().wins += 1;
+                }
+                team_stats.entry(red_team).or_default().losses += 1;
+                team_stats.entry(blue_team).or_default().wins += 1;
+            }
+            _ => {}
+        }
+    }
+
+    let mut final_player_stats = process_stats(player_stats.into_iter());
+    let mut final_team_stats = process_stats(
+        team_stats
+            .into_iter()
+            .map(|(t, s)| (format!("{}, {}", t[0], t[1]), s)),
+    );
+    final_player_stats.sort_by(|a, b| b.pointrate.partial_cmp(&a.pointrate).unwrap());
```

Comment:
`sort_by_key(|stats| cmp::Reverse(stats.pointrate))`
(Also note that default sort is stable. If this is not required `sort_unstable...` should be faster on average)

Reply:
Changing to `sort_unstable_by`. `i64` is only `PartialOrd`, so I followed

https://rust-lang-nursery.github.io/rust-cookbook/algorithms/sorting.html#sort-a-vector-of-floats

Reply:
> `i64` is only `PartialOrd`
Are you sure? https://doc.rust-lang.org/std/primitive.i64.html#impl-Ord-for-i64
Note that all `Ord` classes also implement `PartialOrd` (similarly, `Eq` requires `PartialEq` and `Copy` requires `Clone`)

Reply:
Right, I meant `f64` which the key I chose to use here is (I'm calling it the `pointrate`)

Reply:
I'm surprised they give this advice. I would much rather use [total_cmp](https://doc.rust-lang.org/std/primitive.f64.html#method.total_cmp), which never panics.
This is also what https://doc.rust-lang.org/std/vec/struct.Vec.html#method.sort_floats uses.